### PR TITLE
[Header] Add app block in the header section

### LINKED ISSUE #2284: App Blocks In Header
Waiting for SLT approval to build 

Refined Requirements (mobile only):

- [ ] If aligned center, centers logo automatically between burger icon and left app. Then shrink.

- [ ] If aligned left, shrinks if too horizontal and too many apps.

Stress test prototype with 5-6 different logo sizes.

[DEMO](https://os2-demo.myshopify.com/admin/themes/139242635286/editor?category=gid%3A%2F%2Fshopify%2FOnlineStoreThemeSettingsCategory%2FLogo%3Ftheme_id%3D139242635286%26first_setting_id%3Dlogo)

[STRESS-TESTING](https://www.figma.com/file/KAhNwsSXRi9JPsNRncX38O/Header-flexibility-Phase-2%2F3?node-id=878%3A8838&t=l3vcGHYpjZXNv3xV-1).

### PR DESCRIPTION
**Pull request by ludoboludo**
### PR Summary: 

Add support for app blocks in the header section

Closes #2284 
### Why are these changes introduced?

To enable merchants to add app blocks in their header section, specifically next to their icon sets (search, cart, account). 

### What approach did you take?

Added a block option for apps in the schemas.
We're also forcing the placement to be next to the existing set of icons. 
As well as adding a "safeguard" by limiting the amount of space the block can take -> `4.4rem` by `4.4rem`

### Other considerations

On mobile depending on the amount of app block it will push the logo to the left a little or the merchant has the option to left align their logo on mobile. 

<details>
<summary>Screenshot</summary>

</details>
^ there is an option to align the logo to the left on mobile which will mitigate this issue: 
<details>
<summary>Screenshot</summary>

</details>

Should there be a limit set on the app block ? Eg: limited to 1

### Decision log

| # | Decision | Alternatives | Rationale | Downsides |
|---|---|---|---|---|
| 1 |   |   |   |   |

### Visual impact on existing themes

Nothing on install should change. Only if you add new app blocks.

### Testing steps/scenarios

- [ ] Go into the header section and add an app block. App icon is the one meant for it,
- [ ] But testing with the other available apps can give an idea of what it will look like if you try getting an app that isn't meant for the header 

### Demo links

- [Editor](https://os2-demo.myshopify.com/admin/themes/139275468822/editor)

### Checklist
- [ ] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

**Comment by bakura10** [comment]
Out of curiosity, are you aware of an app offering such an icon block to allow for easier testing of this?

**Comment by ludoboludo** [comment]
Yeah I we kind of share the same concerns @bakura10. We're most likely going to limit the size of the app as well through our CSS as a safeguard but that will be easily overwritten by app code if needed. 
This is the type of thing that we will be iterating on to see how it ends up being used and what situations do we see merchants running into. 

It's to the benefits of apps as well to not break and give a poor experience to merchants. But the merchant could also get confused about which app is meant for the header or not in their app block options. 

As for our testing, we have an app with some theme extensions internally. I can see how it makes it harder for theme partners to try 🤔

**Comment by ludoboludo** [comment]
**Reposting a comment from @bakura10 on a draft PR for the same feature:**

I found a big problem while trying to add app block to the header. In our theme for the header, we have two link_list setting (one for desktop menu and one for drawer).

The issue is that, as soon as we add an app block into the section, Shopify does not allow more than one setting of a given type:

image

We are not ready to trade-off the flexibility here on the menu. This limitation is unfortunately a big blocker for a wider support of this pattern.

Is there a plan to lift this limitation, or at least limit it to some kind of resource. From my understanding, now that we have section groups, because section groups are global (not per template), autofill cannot apply to section groups, so this limitation does not apply.

In all cases, I think this limitation should be first reconsidered before this PR can be merged to Dawn. This would prevent us, theme developers, to offer a feature parity with Dawn to merchant.

**Review by eugenekasimov — COMMENTED** [review]
All looks good to me 👌. Just left one small question 🙂

File: assets/base.css [hunk 1]
```diff
@@ -2309,7 +2309,15 @@ product-info .loading-overlay:not(.hidden) ~ *,
...
 
+.header--has-app {
+  grid-template-columns: auto 2fr 1fr;
+}
+
 @media screen and (min-width: 990px) {
+  .header {
+    grid-template-columns: 1fr auto 1fr;
+  }
+
```
[the remark this answers is not in the document]
Comment (ludoboludo) [reply assets/base.css:2320]: I think having it for the general header when the logo is in the middle is actually useful. And I'm pretty sure it's actually in the latest from `main`. 
Ill rebase to have the latest 👍

**Review by LucasLacerdaUX — APPROVED** [review]
Working as expected and code looks good :D

**Review by eugenekasimov — APPROVED** [review]
🚀

File: assets/base.css [hunk 2]
```diff
@@ -2422,7 +2422,8 @@ product-info .loading-overlay:not(.hidden) ~ *,
     justify-self: start;
   }
 
-  .header--mobile-left {
+  .header--mobile-left,
+  .header--has-app {
```
[the remark this answers is not in the document]
Comment (ludoboludo) [reply assets/base.css]: Yeah I think that's where we trust merchant to make the best decisions for their stores. 👍

**Review by LucasLacerdaUX — DISMISSED** [review]
LGTM :DD

**Comment by bakura10** [comment]
Hi everyone,

@stufreen , I have been told by Ludo that this PR is going to be merged before fixing the issue that prevent us to add this as soon as there are more than one linklist setting. Can the fix in the app block be considered with some high priority? I like the idea of this change, but the fact we have absolutely no way to add this change to our themes because of limitation of app blocks means that we can't offer the same experience as Dawn here. This is really problematic for us, app developers will likely take advantage of this feature, so we would like to be able to offer it as well.

**Comment by stufreen** [comment]
@bakura10 Hi Michaël, I raised the issue and your comment specifically with the App Extensions team and I think they're planning to fix it. No timeline yet, but my understanding is that they are hoping theme dev partners will also add app blocks to headers and this would be a blocker for you and likely some others.

**Comment by bakura10** [comment]
Thanks for the good news ;). Please let me know once this is possible so that we can add this to our themes.

**Review by YoannJailin — APPROVED** [review]

**Review by LucasLacerdaUX — APPROVED** [review]

**Comment by vfonic** [comment]
> **Reposting a comment from @bakura10 on a draft PR for the same feature:**
> 
> I found a big problem while trying to add app block to the header. In our theme for the header, we have two link_list setting (one for desktop menu and one for drawer).
> 
> The issue is that, as soon as we add an app block into the section, Shopify does not allow more than one setting of a given type:
> 
> image
> 
> We are not ready to trade-off the flexibility here on the menu. This limitation is unfortunately a big blocker for a wider support of this pattern.
> 
> Is there a plan to lift this limitation, or at least limit it to some kind of resource. From my understanding, now that we have section groups, because section groups are global (not per template), autofill cannot apply to section groups, so this limitation does not apply.
> 
> In all cases, I think this limitation should be first reconsidered before this PR can be merged to Dawn. This would prevent us, theme developers, to offer a feature parity with Dawn to merchant.



> @ludoboludo Thanks for reposting the comment above. It's not a problem for us since we only have one `link_list` setting, but it's a valid concern. I will pass along to the app extensibility team.


I just hit this issue when trying to upgrade our modified Dawn theme to 9.0.0.
Turns out we have two `link_list` settings in our modified theme.

Any way we could address this?

Thank you!

**Comment by ludoboludo** [comment]
Hey @vfonic, no way to address this at the moment. Ill make sure to relay the feedback to see what could be done/considered on our end 👍

**Pull request merged**